api/agent/agent.go

```diff
@@ -99,8 +99,9 @@ type Agent interface {
 	// immediately before and after the Call is executed, respectively. An error
 	// will be returned if there is an issue executing the call or the error
 	// may be from the call's execution itself (if, say, the container dies,
-	// or the call times out).
-	Submit(Call) error
+	// or the call times out). submitState returned will show the last execution flow
+	// progress Submit() call was able to reach.
+	Submit(Call) (error, submitState)
```

Comment:
could we encode this in the `error` instead? an error type of `submitState` should do. e.g.

```
if s, ok := err.(submitState); ok && s.State() == SubmitStateInitial {
 // ...
}
```

there are a few ways to skin the cat.

Reply:
there is really only one state we are interested in, which is `SubmitStatesWaitSlot`. the rest seem like added complexity. in this case, we should probably just return some kind of `WaitError` type that has a real error on it we can extract, and then we can know that we hit this case by getting a `WaitError` type. then we only have 1 state to handle. beware that embedding errors has the side effect of losing hidden interface implementations (sadness), sorry I'm being intentionally vague to see if you can come up with something better than the rest of us gophers to date >.<

Reply:
no worries, I'll do another pass on this.